File: trunk/xui1.0/backend/PHP/linb.php
<?php
    // input keys
    define('LINB_KEYWORD_CALLBACK', "callback");

    // output keys
    define('LINB_KEYWORD_DATA', "data");
    define('LINB_KEYWORD_ERROR', "error");


    // for json
    if(!function_exists('json_encode')){
        include_once("JSON.php");
        function json_encode($var){
            return (new JSON).encode($var);
        }
        function json_decode($str){
            return (new JSON).decode($str);
        }
    }

    // handle request data
    function linb_getRequestData(){
        $callback = LINB_KEYWORD_CALLBACK;

        $inputData=new stdClass;

        // 1. for "post" request
        // 1.1. form post( a=b$c=d )
        if(count($_POST)>0){
            foreach ($_POST as $k=>$v)
               $inputData->$k = get_magic_quotes_gpc()?stripslashes($v):$v;
        // 1.2. form post( {a:'b',c:'d'} ) or xmlhttp post
        }else{
            $request = file_get_contents('php://input');
            if($request){
                $request = json_decode($request);
                foreach ($request as $k=>$v)
                    $inputData->$k = is_string($v)?get_magic_quotes_gpc()?stripslashes($v):$v:$v;
            }
        }

         // 2. for "get" request
         $request = $_SERVER['QUERY_STRING'];
         // 2.1. get ?a=b$c=d
         if($request){
             if(strstr($request,'=')!==false){
                foreach ($_GET as $k=>$v)
                    $inputData->$k = get_magic_quotes_gpc()?stripslashes($v):$v;
             // 2.2. get ?{a:'b',c:'d'}
             }else{
                $request = json_decode(rawurldecode($request));
                foreach ($request as $k=>$v)
                    $inputData->$k = is_string($v)?get_magic_quotes_gpc()?stripslashes($v):$v:$v;
             }
         }
         return $inputData;
    }
    
    // echo response data, or error info
    function linb_echoResponse($inputData, $outputData, $ok=true){
        $callback = LINB_KEYWORD_CALLBACK;

        $data = LINB_KEYWORD_DATA;
        $err = LINB_KEYWORD_ERROR;

        if(isset($inputData)){
            if(isset($inputData->$callback))
                $callbackV=$inputData->$callback;
            unset($inputData);
        }
                
        $outputDataWrapped=new stdClass;        

        if($ok)
            $outputDataWrapped->$data = $outputData;
        else
            $outputDataWrapped->$err = $outputData;

        $outputDataWrapped=json_encode($outputDataWrapped);

        // wrap result data
 	    // script tag ajax    	    
 	    if(_.isset($callbackV)){
 	        if($callbackV=="window.name"){
 	            $outputDataWrapped="<script type='text' id='json'>".$outputDataWrapped."</script><script type='text/javascript'>window.name=document.getElementById('json').innerHTML;</script>";
 	        }else{
 	            $outputDataWrapped = $callbackV.'('.$outputDataWrapped.')';
 	        }
 	    }
        echo $outputDataWrapped;
    }
    
    /* example
    public function action(){
        $inputData = linb_getRequestData();
        $outputData = null;
        $ok=true;

        // $outputData=...
        // $ok=true;

        linb_echoResponse($inputData, $outputData, $ok);
    }
    */
?>

File: trunk/xui1.0/backend/CSharp/LINB.cs
using System;
using System.Web;
using System.Collections;
using System.Collections.Specialized;
using Jayrock;
using Jayrock.Json;
using Jayrock.Json.Conversion;


namespace SharpLinb
{
    public class LINBException : Exception
    {
        public LINBException()
        {
        }
        public LINBException(string message)
            : base(message)
        {
        }
        public LINBException(string message, Exception inner)
            : base(message, inner)
        {
        }
    }

    public class LINB
    {
        public const string ERR = "error";
        public const string DATA = "data";
        public const string CALLBACK = "callback";


        public static Hashtable getRequestData(HttpContext context)
        {
            Hashtable hRequest = new Hashtable();

            //get request data from POST
            if (context.Request.Form.HasKeys())
            {
                LINB.merge(context.Request.Form, hRequest);
            }
            else
            {
                String sRawPostData = context.Request.Form[null];
                if (sRawPostData != null && sRawPostData.Trim().Length > 0)
                {
                    JsonObject obj = (JsonObject)JsonConvert.Import(sRawPostData);
                    LINB.merge(obj, hRequest);
                }

            }

            //get request data from GET
            if (context.Request.QueryString.HasKeys())
            {
                LINB.merge(context.Request.QueryString, hRequest);
            }
            else
            {
                String sQueryString = context.Request.QueryString[null];
                if (sQueryString != null && sQueryString.Trim().Length > 0)
                {
                    sQueryString = context.Server.UrlDecode(sQueryString);
                    JsonObject obj = (JsonObject)JsonConvert.Import(sQueryString);
                    LINB.merge(obj, hRequest);
                }
            }

            return hRequest;
        }


        public static void echoResponse(HttpContext context, Hashtable inputData, Object outputData, Boolean ok)
        {
            context.Response.ClearContent();
            string strOut = "";

            Hashtable hResponse = new Hashtable();

            if (ok)
            {
                hResponse.Add(DATA, outputData);
            }
            else
            {
                hResponse.Add(ERR, outputData);
            }
            Object oCallback = inputData[CALLBACK];
            //adjust out string
            if (oCallback != null)
            {
                string sCallback = "" + oCallback;
                if ("window.name".Equals(sCallback))
                {
                    context.Response.ContentType = "text/html";
                    strOut = "<script type='text' id='json'>"
                        + JsonConvert.ExportToString(hResponse)
                        + "</script><script type='text/javascript'>window.name=document.getElementById('json').innerHTML;</script>";
                }
                else
                {
                    context.Response.ContentType = "text/plain";
                    strOut = sCallback + "(" + JsonConvert.ExportToString(hResponse) + ")";
                }
            }
            else
            {
                context.Response.ContentType = "text/plain";
                strOut = JsonConvert.ExportToString(hResponse);
            }


            if (inputData != null)
                inputData.Clear();
            context.Response.Write(strOut);
            context.Response.End();
        }

        private static void merge(NameValueCollection col, Hashtable hash)
        {
            foreach (string key in col)
            {
                hash[key] = col[key];
            }
        }
        private static Hashtable JsonObjectToHash(JsonObject jo)
        {
            Hashtable hOut = new Hashtable();
            IDictionaryEnumerator i = jo.GetEnumerator();
            i.MoveNext();
            do
            {
                hOut[i.Key.ToString()] = i.Value.ToString();
            } while (i.MoveNext());
            return hOut;
        }
        private static void merge(JsonObject col, Hashtable hash)
        {
            IDictionaryEnumerator i = col.GetEnumerator();
            i.MoveNext();
            do
            {
                if (i.Value is JsonObject)
                {
                    hash[i.Key] = JsonObjectToHash((JsonObject)i.Value);
                }
                else
                    hash[i.Key] = i.Value;
            } while (i.MoveNext());
        }
    }
}
